Sources/Controllers/MessagesViewController.swift

```diff
@@ -303,11 +331,40 @@ fileprivate extension MessagesViewController {
     /// is larger than the required offset for the MessagesCollectionView
     ///
     /// - Returns: The safeAreaInsets.bottom if its an iPhoneX, else 0
-    fileprivate var iPhoneXBottomInset: CGFloat {
+    private var iPhoneXBottomInset: CGFloat {
         if #available(iOS 11.0, *) {
             guard UIScreen.main.nativeBounds.height == 2436 else { return 0 }
             return view.safeAreaInsets.bottom
         }
         return 0
     }
+
+    private func calculateBottomInset(with keyboardEndFrame: CGRect) -> CGFloat? {
+        // Happens when the user drags undocked keyboard
+        guard keyboardEndFrame.height > 0 else { return nil }
+
+        let keyboardFrame = view.convert(keyboardEndFrame, from: UIScreen.main.coordinateSpace)
+        let intersection = keyboardFrame.intersection(view.bounds)
+
+        let keyboardHeight: CGFloat
+
+        // check if keyboard covers our view
+        if intersection.height > 0 {
+            // We can't use `intersection.height` as `keyboardHeight` because the keyboard can be in undocked mode
+            keyboardHeight = view.bounds.maxY - keyboardFrame.minY
+        } else {
+            keyboardHeight = 0
+        }
+
+        return keyboardHeight - iPhoneXBottomInset
```

Comment:
This all could be cleaned up into a smaller function.

`return intersection.height > 0 ? (view.bounds.maxY - keyboardFrame.minY - iPhoneXBottomInset ) : -iPhoneXBottomInset`

Reply:
I didn't use because I wanted add this comment: "We can't use `intersection.height` as `keyboardHeight` because the keyboard can be in undocked mode".
Otherwise this code might confuse - why not just use `intersection.height`?